File: BACK-END/KuentecoApp/src/main/java/org/kuenteco/backend/dto/logic/transaction/kuenteco/NewTransactionDTO.java
package org.kuenteco.backend.dto.logic.transaction.kuenteco;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;
import java.math.BigDecimal;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.kuenteco.backend.entity.extra.DescriptionTransaction;
import org.kuenteco.backend.validation.ValidCurrency;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class NewTransactionDTO {
    @Positive(message = "El ID de categoría debe ser un número positivo")
    private Integer categoryId;

    @Positive(message = "El ID de presupuesto debe ser un número positivo")
    private Integer budgetId;

    @Positive(message = "El ID de deuda debe ser un número positivo")
    private Integer debtId;

    @NotBlank(message = "El nombre de la transacción es requerido")
    @Size(min = 3, max = 255, message = "El nombre debe tener entre 3 y 255 caracteres")
    private String name;

    @Valid
    @NotNull(message = "La descripción de la transacción es requerida")
    private DescriptionTransaction description;

    @NotNull(message = "El monto es requerido")
    @ValidCurrency(
            min = 0.01,
            max = 999999999.99,
            message = "El monto debe ser un valor monetario válido")
    private BigDecimal amount;
}

File: FRONT-END/lib/dto/transaction/kuenteco/new_transaction_dto.dart
import '../../extra/description_transaction_extra.dart';

class NewTransactionDTO {
  final int? categoryId;
  final int? budgetId;
  final int? debtId;
  final String? name;
  final DescriptionTransaction description;
  final double amount;

  NewTransactionDTO({
    this.categoryId,
    this.budgetId,
    this.debtId,
    this.name,
    required this.description,
    required this.amount,
  });

  factory NewTransactionDTO.fromJson(Map<String, dynamic> json) {
    return NewTransactionDTO(
      categoryId: json['categoryId'],
      budgetId: json['budgetId'],
      debtId: json['debtId'],
      name: json['name'],
      description: DescriptionTransaction.fromJson(json['description']),
      amount: (json['amount'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'categoryId': categoryId,
      'budgetId': budgetId,
      'debtId': debtId,
      'name': name,
      'description': description.toJson(),
      'amount': amount,
    };
  }
}
